client/tso_stream.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// TSO Stream Builder
+
+type tsoStreamBuilder interface {
+	build(context.Context, context.CancelFunc, time.Duration) (tsoStream, error)
+}
+
+type pdTSOStreamBuilder struct {
+	client pdpb.PDClient
+}
+
+func (b *pdTSOStreamBuilder) build(ctx context.Context, cancel context.CancelFunc, timeout time.Duration) (tsoStream, error) {
+	done := make(chan struct{})
+	// TODO: we need to handle a conner case that this goroutine is timeout while the stream is successfully created.
+	go checkStreamTimeout(ctx, cancel, done, timeout)
+	stream, err := b.client.Tso(ctx)
+	done <- struct{}{}
+	if err == nil {
+		return &pdTSOStream{stream: stream}, nil
+	}
+	return nil, err
+}
+
+type tsoTSOStreamBuilder struct {
+	client tsopb.TSOClient
+}
+
+func (b *tsoTSOStreamBuilder) build(ctx context.Context, cancel context.CancelFunc, timeout time.Duration) (tsoStream, error) {
+	done := make(chan struct{})
+	// TODO: we need to handle a conner case that this goroutine is timeout while the stream is successfully created.
+	go checkStreamTimeout(ctx, cancel, done, timeout)
+	stream, err := b.client.Tso(ctx)
+	done <- struct{}{}
+	if err == nil {
+		return &tsoTSOStream{stream: stream}, nil
+	}
+	return nil, err
+}
+
+func checkStreamTimeout(ctx context.Context, cancel context.CancelFunc, done chan struct{}, timeout time.Duration) {
+	select {
+	case <-done:
+		return
+	case <-time.After(timeout):
+		cancel()
+	case <-ctx.Done():
+	}
+	<-done
+}
+
+// TSO Stream
+
+type tsoStream interface {
+	// processRequests processes TSO requests in streaming mode to get timestamps
+	processRequests(clusterID uint64, dcLocation string, requests []*tsoRequest,
+		batchStartTime time.Time) (physical, logical int64, suffixBits uint32, err error)
+}
+
+type pdTSOStream struct {
+	stream pdpb.PD_TsoClient
+}
+
+func (s *pdTSOStream) processRequests(clusterID uint64, dcLocation string, requests []*tsoRequest,
+	batchStartTime time.Time) (physical, logical int64, suffixBits uint32, err error) {
+	start := time.Now()
+	count := int64(len(requests))
+	req := &pdpb.TsoRequest{
+		Header: &pdpb.RequestHeader{
+			ClusterId: clusterID,
+		},
+		Count:      uint32(count),
+		DcLocation: dcLocation,
+	}
+
+	if err = s.stream.Send(req); err != nil {
+		err = errors.WithStack(err)
+		return
+	}
+	tsoBatchSendLatency.Observe(float64(time.Since(batchStartTime)))
+	resp, err := s.stream.Recv()
+	if err != nil {
+		err = errors.WithStack(err)
+		return
+	}
+	requestDurationTSO.Observe(time.Since(start).Seconds())
+	tsoBatchSize.Observe(float64(count))
+
+	if resp.GetCount() != uint32(count) {
+		err = errors.WithStack(errTSOLength)
+		return
+	}
+
+	physical, logical, suffixBits = resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical(), resp.GetTimestamp().GetSuffixBits()
+	return
+}
+
+type tsoTSOStream struct {
+	stream tsopb.TSO_TsoClient
+}
+
+func (s *tsoTSOStream) processRequests(clusterID uint64, dcLocation string, requests []*tsoRequest,
```

Comment:
this two function is so similiar

Reply:
It's not easy to remove duplicate code now, because the request in stream.Send(), the stream, the response from stream.Recv() are in different types from pdpb and tsopb.  We have used several design patterns, including template, Abstract Factory, Builder as you can see in base_client.go and tso_stream.go, so that we can share the service discovery code and tso batching/forwarding/async framework. These two functions are closest to the protobuf code and only different in types so it is even not possible to dedup further.